snapcraft/tests/plugins/python/test_python_finder.py

```diff
@@ -0,0 +1,114 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2017 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import os
+import re
+
+from testtools.matchers import (
+    Equals,
+    MatchesRegex,
+)
+
+from snapcraft.plugins._python import _python_finder
+
+from snapcraft import (
+    tests
+)
+
+
+class GetPythonCommandTestCase(tests.TestCase):
+
+    def test_staged(self):
+        """get_python_command should support staged python"""
+
+        stage_dir = 'stage_dir'
+        install_dir = 'install_dir'
+
+        # Create staged binary
+        _create_python_binary(stage_dir)
+
+        python_command = _python_finder.get_python_command(
+            'python2', stage_dir, install_dir)
+        self.assertThat(
+            python_command, Equals(os.path.join(
+                stage_dir, 'usr', 'bin', 'python2')))
+
+    def test_installed(self):
+        """get_python_command should support installed python"""
+
+        stage_dir = 'stage_dir'
+        install_dir = 'install_dir'
+
+        # Create installed binary
+        _create_python_binary(install_dir)
+
+        python_command = _python_finder.get_python_command(
+            'python2', stage_dir, install_dir)
+        self.assertThat(
+            python_command, Equals(os.path.join(
+                install_dir, 'usr', 'bin', 'python2')))
+
+    def test_staged_and_installed(self):
+        """get_python_command should prefer staged python over installed"""
+
+        stage_dir = 'stage_dir'
+        install_dir = 'install_dir'
+
+        # Create both staged and installed binaries
+        _create_python_binary(stage_dir)
+        _create_python_binary(install_dir)
+
+        python_command = _python_finder.get_python_command(
+            'python2', stage_dir, install_dir)
+        self.assertThat(
+            python_command, Equals(os.path.join(
+                stage_dir, 'usr', 'bin', 'python2')))
+
+
+class GetPythonHeadersTestCase(tests.TestCase):
```

Comment:
no test for when both exist and staged is preferred?
no test for when '' is returned?

Reply:
The first test actually exists, because there's no way to NOT have the headers installed on the host for these tests. Which also means I don't know of a way to test the `''` case. I can probably mock `os.exists` or something, though, but that sounds a little anti-Leo. Thoughts?

Reply:
I vote for anti-Leo, mocking os.exists.
But, now I found a middle-ground, like in my proposal to mock pip. You can mock os.exists only if it receives the argument you are sending, and use te real os.exists in all the other cases. I'm starting to like that pattern, please let me know your thoughts because I think I'll start using it everywhere.

Reply:
Anti-Leo, done. That should be completely covered in tests, now.